ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -672,22 +672,46 @@ private IndexValue markAsDeleted(StoreKey id, FileSpan fileSpan, MessageInfo inf
    * @throws StoreException if there is any problem writing the index record
    */
   IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, long operationTimeMs) throws StoreException {
+    return markAsPermanent(id, fileSpan, null, operationTimeMs);
+  }
+
+  /**
+   * Marks a blob as permanent
+   * @param id the {@link StoreKey} of the blob
+   * @param fileSpan the file span represented by this entry in the log
+   * @param operationTimeMs the time of the update operation
+   * @param info this needs to be non-null in the case of recovery. Can be {@code null} otherwise. Used if the PUT
+   *             record could not be found
+   * @return the {@link IndexValue} of the ttl update record
+   * @throws StoreException if there is any problem writing the index record
+   */
+  private IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, MessageInfo info, long operationTimeMs)
+      throws StoreException {
     validateFileSpan(fileSpan, true);
     IndexValue value = findKey(id);
-    if (value == null) {
+    if (value == null && info == null) {
       throw new StoreException("Id " + id + " not present in index " + dataDir, StoreErrorCodes.ID_Not_Found);
-    } else if (value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+    } else if (value != null && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
       throw new StoreException("Id " + id + " deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
-    } else if (value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
+    } else if (value != null && value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
       throw new StoreException("TTL of " + id + " already updated in index" + dataDir, StoreErrorCodes.Already_Updated);
     }
     long size = fileSpan.getEndOffset().getOffset() - fileSpan.getStartOffset().getOffset();
-    IndexValue newValue =
-        new IndexValue(value.getSize(), value.getOffset(), value.getFlags(), Utils.Infinite_Time, operationTimeMs,
-            value.getAccountId(), value.getContainerId());
+    IndexValue newValue;
+    if (value == null) {
+      // It is possible that the PUT has been cleaned by compaction
```

Comment:
What does it mean to mark a blob as permanent (by adding an entry to the index) when the PUT was cleaned up by compaction?

Reply:
Semantically, it means nothing (it would be ok to even lose it) - this is to make sure that we don't lose history (everything in the log is represented in the index) and to not crash recovery. An alternate would be to catch a `StoreException` with error code `ID_Deleted` on `markAsPermanent()` and continue but then the index would have a hole in its representation of the log.

Reply:
This is a bit confusing. Could you clarify what you mean when you say "the PUT has been cleaned by compaction"? Do you mean cleaned up only in the index and not in the Log (somehow)?

What recovery finds in the Log in the scenario you are addressing is a TTL_UPDATE record, and not a PUT record. How do you even know that the Log has a PUT record and that it was not cleaned up?

I want to make sure that we do not end up in a situation where we mark a blob as permanent in the index (thereby marking it as "present") when the record is absent in the Log. Is that guaranteed with this change?

Reply:
Yes, nothing in the index is "extra" from what is in the log. Here is a scenario when the code in this change will actually execute. 

1. Put blob A with a TTL (this is in log segment 0)
2. Mark A as permanent (this is in log segment 1)
3. Delete A
(state at this time: A has a PUT, TTL_UPDATE and DELETE entry in both log and index)
4. Compaction runs on segment 0 and cleans the PUT record of A (because it is deleted)
(state at this time: A has a TTL_UPDATE and DELETE entry in both log and index)
5. The index is deleted and the store restarted
6. Recovery runs and finds the TTL_UPDATE of A. There is no corresponding PUT record (nothing in the log, so it follows that there is nothing in the index). 
7. Recovery completes after exercising new code
(state at this time: A has a TTL_UPDATE and DELETE entry in both log and index i.e. exactly what it was before the index was deleted)

Without this code, step 6 would have failed.

Reply:
it is just a little counter-intuitive that this would continue to add to the index and "make it permanent" when it knows that the blob is irretrievable (because it does not exist any more). It expects a DELETE to come in to make things consistent).

`markAsDeleted()` does not have this problem, as it does not require an older PUT record.

Reply:
Could you elaborate? You can make the same argument for DELETE right? DELETE _does_ require an older PUT and uses the same mechanism to account for PUTs not being present during recovery. I don't see a difference.

Reply:
and the DELETE is _surely_ present - otherwise compaction wouldn't have cleaned up the PUT

Reply:
To me, `markAsDeleted()`  is different in that you are simply marking a key as deleted, and it serves to return a `Blob_Deleted` error. To return that error, you do not need the original record.

`markAsPermanent()` marks a blob as permanent (or in the general case, "updates" the record), and in this case that record does not exist. It is counter intuitive in that you haven't made a record permanent really. It is supposed to serve to actually return the record, but you never can. And it works because, outside of this method, all callers are now required to ensure that it is going to mark this record as deleted in the future. It is all in the caller's logic, so I don't find this method by itself very intuitive. The method when read does not make sense by itself - in this scenario, *at the time it marks the record as permanent, there is no such record*.

I do not think the index needs to *exactly* reflect what is in the Log, record by record, why is that a requirement? Index is an auxiliary data structure that needs to be performant and serve correct results. I would prefer alternatives, if possible and easy - such as filtering out the entries that are retrieved by recovery (à la `eliminateDuplicates()`) before proceeding to updating them.

Reply:
As discussed, made a change to "expect" deletes when TTL updates are applied without a corresponding PUT record. Recovery will fail if the delete is not seen eventually (added a test)